US-65 Edit/ShareFolderActivity

techbykim opened the pull request:
Taiga Task:
- https://tree.taiga.io/project/dzemlin-poste/us/65?milestone=364858

Screenshots:
- ![image](https://github.com/Appyo-Poste/Poste/assets/57041732/e24c0317-3ef6-4f7b-8027-849dde5233a3)
- ![image](https://github.com/Appyo-Poste/Poste/assets/57041732/124af789-25d1-4d48-a0d6-cbfb1b18d222)

techbykim committed:
task 70 refactor edit folder activity

techbykim committed:
task 79; remove existing code for EditFolderActivity and update to v2

techbykim committed:
init shared folder activity xml v2

techbykim committed:
task 80; remove activity_shared_folder

techbykim committed:
task 80

techbykim committed:
task 71; init Shared_Folder_v2

techbykim committed:
owner -> full_access

techbykim committed:
resolve build issue

techbykim committed:
Merge branch 'dev' into US65-Edit/ShareFolderActivity

techbykim committed:
resolve build issues

techbykim committed:
Merge branch 'dev' into US65-Edit/ShareFolderActivity

dzemlin commented on app/src/main/res/layout/activity_shared_folder_v2.xml:
--Non-blocking Suggestion--
If the old version is being deleted in this same pull, can we drop the _v2 from the end of the file name?

dzemlin commented on app/src/main/java/com/example/poste/api/poste/models/FolderAccess.java:
```diff
@@ -16,9 +16,9 @@ public enum FolderAccess {
+            case 1: return VIEWER;
+            case 2: return EDITOR;
+            case 3: return FULL_ACCESS;
             default: return NONE;
```
--Non-blocking Question--
I Don't think this hurts anything, but I thought Folder access no longer used this class. I thought it was changed to just a simple string or something. maybe I'm wrong?
either way no harm in this code.

mccodav1 replied:
These are just enums, so I don't see the harm here. It might be more clear to use the models or utils classes we've moved functionality to, though.

dzemlin commented on app/src/main/java/com/example/poste/activities/DashboardActivity.java:
```diff
@@ -273,4 +273,72 @@ public void onClick(DialogInterface dialog, int which) {
+        }
+        return true;
+    }
 }
```
--Non-blocking question--
Would you be against having this context menu only pop up on a long-click?
and a single tap just opens the file?
If so I can adapt that to the MVP.

dzemlin reviewed (approved):
Ok... I promise I had a lot more comments at first but the more I thought about them the more I realized you were right.
This PR has my approval.
Just please read the few comment i left in for suggestions and respond, as they will be relevant to my branch.
Great work!

mccodav1 reviewed (changes requested):
Lots of work here, thank you!

A couple things I noticed:

- When attempting to edit a folder from the dashboard, the Edit Folder menu comes up with an option to select name. However, when trying to change the name, the backend does not edit the folder. The server logs show:
```
[03/Nov/2023 22:35:20] "GET /api/folders/id/-1 HTTP/1.1" 302 0
[03/Nov/2023 22:35:20] "GET /login/?next=/api/folders/id/-1 HTTP/1.1" 200 4191
Unauthorized: /api/data/
[03/Nov/2023 22:35:21] "GET /api/data/ HTTP/1.1" 401 59
```

This also crashes the app. In Logcat:
```
FATAL EXCEPTION: main
                                                                                                    Process: com.example.poste, PID: 19074
                                                                                                    java.lang.RuntimeException: com.example.poste.api.poste.exceptions.MalformedResponseException: The API's response was malformed and could not be parsed
                                                                                                    	at com.example.poste.activities.EditFolderActivity_v2.lambda$onCreate$0$com-example-poste-activities-EditFolderActivity_v2(EditFolderActivity_v2.java:64)
...                                                                                              	
```

- When attempting to share a folder from dashboard, Share Folder menu comes up with two different dropdowns. Both dropdowns display permission choices: no access, view access, and manage access. This also gives an odd request to the backend with a malformed response

Tomorrow I may be able to pick through the code a bit to see what is happening, but functionally this is causing some odd behaviors

mccodav1 commented on app/src/main/java/com/example/poste/activities/Shared_Folder_v2.java:
```diff
@@ -45,34 +45,42 @@ protected void onCreate(Bundle savedInstanceState) {
+            String folderSelectedAccessValue = folderSpinner.getSelectedItem().toString();
+            String permissionsSelectedAccessValue = permissionsSpinner.getSelectedItem().toString();
             try {
                 Folder targetFolder = API.getFolderById(folderId);
```
These API calls have been deprecated in favor of the new HTTP helper methods / retrofit client; the other activities will have examples. See com.example.poste.http package for more context

mccodav1 commented on app/src/main/java/com/example/poste/activities/Shared_Folder_v2.java:
```diff
@@ -45,34 +45,42 @@ protected void onCreate(Bundle savedInstanceState) {
-        spinner.setAdapter(spinnerAdapter);
+        Spinner folderSpinner = findViewById(R.id.share_folder_spinner);
+        Spinner permissionsSpinner = findViewById(R.id.share_folder_permissions_spinner);
+        ArrayAdapter<CharSequence> folderSpinnerAdapter = ArrayAdapter.createFromResource(this, R.array.spinner_values, android.R.layout.simple_spinner_item);
```
I think this is why both spinners populate with permissions -- they seem to be built from the same resource

techbykim committed:
task 186; make edit work

techbykim committed:
task 188; make update work

techbykim commented:
Mostly working except for one issue with the share permission tokens which I describe here: https://www.loom.com/share/d24511ecc93b49fe9475a009e87b4b0d

techbykim committed:
working; needed '/' and to send to dashboard

techbykim commented:
It is working now; thanks @mccodav1 for realizing the issue with the trailing `/`

dzemlin commented:
Ok. So I saw in your video it would not open the edit post page from the folder view page. 
Do not worry about fixing that, it is already fixed in my branch as it was a problem in FolderViewActivity.
when I merged your branch with mine it works just fine. 
So you may want to take the FolderViewActivity.java file from my branch (us128...) and copy that into yours for testing.

There a 2 other errors that happen though when you are actually able to get into the EditPost view.

1. the data does not populate into the fields. but I already found the issue and fix (see my comment I just added to EditPostActivity.java)

2. it has an error when you try to save the post edits you just made. It does not record the changes and sends an error toast message. I think something is wrong with the API call it is making.

mccodav1 committed:
Add some conversions / string values to enums

Kim -- This is kinda messy and probably defeats the purposes of using an enum
in the first place. I'm just so used to using constants rather than enums...
and this is just how I got this working with the backend the way it is. Originally
I wanted to ensure we sent a string so we didnt have to change the backend, but the
backend ended up requiring changes anyway so... idk. Oh well.

This is tested working with the backend changes I made a PR for back to your branch.

Feel free to take / leave / modify as you see fit. Not trying to step on toes, just
trying to assist, so if you want to ignore these PR please feel free.

techbykim committed:
Add some conversions / string values to enums #18 from Appyo-Poste/US65-debug
